### PR TITLE
[WiP] - inform administrator about plugins incompatibilities and missing plugins

### PR DESCRIPTION
**Pull request by ndeloof**
show alert message on "manage" view and in logs when any plugin doesn't get it's dependency minimal versions installed.

Maybe I could even use same hack the "please wait jenkins to start" message uses to show this at startup, until some user acknowledge and switch to standard UI.

### CODE REVIEW

File: core/src/main/java/hudson/PluginManager.java [hunk 1]
```diff
@@ -327,6 +322,35 @@ protected void reactOnCycle(PluginWrapper q, List<PluginWrapper> cycle)
...
                             });
 
+
+                            g.followedBy().attains(PLUGINS_LISTED).add("Checking plugin dependencies", new Executable() {
+                                @Override
+                                public void run(Reactor reactor) throws Exception {
+                                    for (PluginWrapper plugin: getPlugins()) {
+                                        String core = plugin.getManifest().getMainAttributes().getValue("Jenkins-Version");
+                                        if (core != null && Jenkins.getVersion().compareTo(new VersionNumber(core)) < 0) {
+                                            String s = "Plugin " + plugin.getShortName() + " required jenkins " + core + " or later.";
```
Comment (daniel-beck) [inline core/src/main/java/hudson/PluginManager.java]: * `requires`
* Unlocalizable

File: core/src/main/java/hudson/PluginManager.java [hunk 2]
```diff
@@ -327,6 +322,35 @@ protected void reactOnCycle(PluginWrapper q, List<PluginWrapper> cycle)
...
+                                            LOGGER.severe(s);
+                                            NOTICE.add(s);
+                                        }
+
+                                        for (Dependency dep : plugin.getDependencies()) {
+                                            if (dep.optional) continue;
+                                            PluginWrapper p = getPlugin(dep.shortName);
+                                            if (p == null) {
+                                                String s = "Plugin " + plugin.getShortName() + " is missing required dependency " + p.getShortName();
```
Comment (daniel-beck) [inline core/src/main/java/hudson/PluginManager.java]: Users are generally not shown plugin short names and shouldn't here either (at least not *only* short names).

**Comment by daniel-beck** [comment]
This appears to resolve (at least partially) JENKINS-23150, so should mention that.

Wouldn't it make more sense to make at least the UI parts a standalone component that could even be moved into a (bundled) plugin to enable core-independent updates, rather than integrate with plugin manager?

This should also show plugins that failed to load (even if they're not dependencies of anything). See JENKINS-21485 and issues linked from there.

Something else that comes up occasionally is implied dependencies to detached plugins from plugins based on older cores. There should be a way for administrators to these "possible dependency problems" with some explanation why they may matter.

I don't think this should be made into a `HudsonIsLoading` type screen, as that has no authentication support and would only confuse users of a multi-user instance. If you're concerned about visibility of an admin monitor, use https://github.com/daniel-beck/jenkins-visible-administrative-monitors-plugin.

File: core/src/main/java/hudson/PluginManager.java [hunk 3]
```diff
@@ -327,6 +322,35 @@ protected void reactOnCycle(PluginWrapper q, List<PluginWrapper> cycle)
...
+                                                LOGGER.severe(s);
+                                                NOTICE.add(s);
+                                            } else if (p.getVersionNumber().compareTo(new VersionNumber(dep.version)) < 0) {
+                                                String s = "Plugin " + plugin.getShortName() + " require " + p.getShortName() + " " + dep.version + " or later";
```
Comment (daniel-beck) [inline core/src/main/java/hudson/PluginManager.java]: `requires`

**Comment by ndeloof** [comment]
@daniel-beck thanks for the review. Plan to move visible-administrative-monitors-plugin to jenkinsci org ?

**Comment by daniel-beck** [comment]
Yes, whenever I get around to writing a config UI for it rather than control via System properties, and figuring out how to release things.

**Comment by jglick** [comment]
-0. I would take a harder line and implement [JENKINS-21486](https://issues.jenkins-ci.org/browse/JENKINS-21486)—simply refuse to load plugins not matching their dependencies, unless a system property is set. There is already an administrative warning shown if some plugins failed to load.

In other words, final enforcement of dependencies should be strict and done during startup with no UI to speak of. We should rather concentrate effort on making Jenkins not get itself into this position to begin with: do not offer incompatible plugins in _Updates_ or _Available_ (show them but with a disabled checkbox, or just hide them, since the update center is expected to be doing this filtering to begin with); do not allow a plugin to be disabled if it has enabled dependencies; verify dependencies in _Upload_ from the _Advanced_ tab; etc. There will always be some corner cases where the UI does not save you, for example because you downgraded core, or added plugins directly to the filesystem, but most people should not encounter those situations.

**Comment by daniel-beck** [comment]
> with no UI to speak of.

Disabling plugins without an admin UI will lead to data loss due to form submissions when a missing plugin wasn't noticed at first.

> added plugins directly to the filesystem, but most people should not encounter those situations.

Search for 'chef' in Jira. It's insane.

**Comment by jglick** [comment]
> when a missing plugin wasn't noticed at first

Well you would get an administrative warning just as this PR proposes. (I like your idea of pushing warnings onto other Jenkins pages when the authenticated user is an administrator; I see no reason to limit them to `/manage`.)

Yes I know people deploy instances including plugins via Chef (or whatever), but I do not see that as a problem at all. With my proposed change, they will get immediate feedback that they screwed up the transitive dependency list, because the plugin will not be loaded at all, rather than finding this out later because there is only a warning but the plugin is still loaded, or only because (as in the current state) certain plugin functions throw mysterious `LinkageError`s.

**Comment by daniel-beck** [comment]
> In other words, final enforcement of dependencies should be strict and done during startup with no UI to speak of. 

You mean like this?

https://issues.jenkins-ci.org/browse/JENKINS-27130

**Comment by jglick** [comment]
> You mean like this?

Precisely the opposite: enforcing dependencies would mean a simple message in the log file and an admin warning, rather than a messy `NoClassDefFoundError` stack trace.

**Comment by daniel-beck** [comment]
@jglick Your proposed solution does not cover implied dependencies to detached plugins either, or does it? (E.g. [JENKINS-27227](https://issues.jenkins-ci.org/browse/JENKINS-27227))

**Comment by jglick** [comment]
> Your proposed solution does not cover implied dependencies to detached plugins either, or does it?

I think it does. In that case the old version of the `tap` plugin would get an injected dependency on `junit` (by `ClassicPluginStrategy`), so if `junit` were disabled, Jenkins would refuse to load `tap` either.

It just needs to be ensured that the resulting admin warning indicates that the `tap` plugin did not get loaded, and that it has an unsatisfied dependency.

In this case it is unclear to me what user mistake caused the breakage. Just updating Jenkins core should not have made that happen, because then the bundled `junit` plugin would still be enabled. Perhaps the user updated and _then_ disabled `junit`. So under my suggested complex of fixes, that disabling would have been blocked by the _Plugin Manager_ UI before it got to this point.

**Comment by daniel-beck** [comment]
In other words, plugins that don't even use detached plugins but have ancient parent versions will block disabling these plugins for no reason? Not sure this is a great idea.

**Comment by jglick** [comment]
> plugins that don't even use detached plugins but have ancient parent versions will block disabling these plugins for no reason

Well there is a reason: we have no idea whether they actually refer to the now-split classes, so we must assume they do. If you think you know better, you can run Jenkins with a system property to suppress dependency enforcement and deal with the consequences.

**Comment by daniel-beck** [comment]
This would be a reasonable compromise.

Introducing that will need another wiki page dedicated to it, good error messages, and big fat warnings on the users list and changelog... probably a Jira banner... this will break a lot of existing installs.

**Comment by jtnord** [comment]
is there an agreement on what should be happening to this PR?
Closed in favour of a different approach or updated accordingly?

**Comment by ndeloof** [comment]
I keep this one open as a reminder to create a fresh new one to concile all comments

**Comment by jglick** [comment]
I do not think there is a clear consensus yet. My current feeling is that this PR is OK in isolation but not really sufficient.

**Comment by ndeloof** [comment]
Closing and will consider an alternate impl

**Comment by daniel-beck** [comment]
@jglick What about a mechanism that allows a plugin depending on an old core to declare a dependency on a detached plugin, which gets ignored on cores that don't know about that plugin (i.e. for those in which it still is core functionality), but making it a hard error on cores that do? This would keep plugins compatible with old Jenkinses, while ensuring valid configurations on newer ones.

**Comment by jglick** [comment]
I think that was already covered in [JENKINS-28942](https://issues.jenkins-ci.org/browse/JENKINS-28942).

**Comment by fbelzunc** [comment]
@ndeloof I continue with your work here #2098

**Pull request closed without merging**